core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -199,41 +241,120 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
     removeExecutor(executorRemoved.executorId)
   }
 
+  private def killExecutor(executorId: String, timeout: Long): Unit = {
+    logWarning(s"Removing executor $executorId with no recent heartbeats: " +
+      s"${timeout} ms exceeds timeout $executorTimeoutMs ms")
+    killExecutorThread.submit(new Runnable {
+      override def run(): Unit = Utils.tryLogNonFatalError {
+        // Note: we want to get an executor back after expiring this one,
+        // so do not simply call `sc.killExecutor` here (SPARK-8119)
+        sc.killAndReplaceExecutor(executorId)
+        // SPARK-27348: in case of the executors which are not gracefully shut down,
+        // we should remove lost executors from CoarseGrainedSchedulerBackend manually
+        // here to guarantee two things:
+        // 1) explicitly remove executor information from CoarseGrainedSchedulerBackend for
+        //    a lost executor instead of waiting for disconnect message
+        // 2) call scheduler.executorLost() underlying to fail any tasks assigned to
+        //    those executors to avoid app hang
+        sc.schedulerBackend match {
+          case backend: CoarseGrainedSchedulerBackend =>
+            // TODO (SPARK-39984): Update causedByApp when we have a hanging task detector
+            backend.driverEndpoint.send(RemoveExecutor(executorId,
+              ExecutorProcessLost(
+                s"Executor heartbeat timed out after ${timeout} ms")))
+          // LocalSchedulerBackend is used locally and only has one single executor
+          case _: LocalSchedulerBackend =>
+
+          case other => throw new UnsupportedOperationException(
+            s"Unknown scheduler backend: ${other.getClass}")
+        }
+      }
+    })
+  }
+
+  private def isStandalone(): Boolean = {
+    sc.schedulerBackend match {
+      case backend: StandaloneSchedulerBackend => true
+      case _ => false
+    }
```

Comment:
Hi @Ngone51, thank you for your review! 

Because `schedulerBackend` is possible to be None, I used pattern matching instead of `isInstanceOf`. @mridulm and I have discussed the issues above. I will tag you in these two discussion threads.

Reply:
Applied the suggestions.